src/driver/amdxdna/amdxdna_gem.c

```diff
@@ -342,59 +435,40 @@ int amdxdna_drm_create_bo_ioctl(struct drm_device *dev, void *data, struct drm_f
 	if (args->flags || !args->size)
 		return -EINVAL;
 
-	XDNA_DBG(xdna, "type %d vaddr 0x%llx size 0x%llx flags 0x%llx",
+	XDNA_DBG(xdna, "BO arg type %d vaddr 0x%llx size 0x%llx flags 0x%llx",
 		 args->type, args->vaddr, args->size, args->flags);
-	mutex_lock(&client->mm_lock);
 	switch (args->type) {
 	case AMDXDNA_BO_SHMEM:
 		abo = amdxdna_drm_alloc_shmem(dev, args, filp);
 		break;
 	case AMDXDNA_BO_DEV_HEAP:
-		if (client->dev_heap != AMDXDNA_INVALID_BO_HANDLE) {
-			XDNA_DBG(client->xdna, "dev heap is already created");
-			ret = -EINVAL;
-			goto err_unlock;
-		}
-
 		abo = amdxdna_drm_create_dev_heap(dev, args, filp);
 		break;
 	case AMDXDNA_BO_DEV:
-		abo = amdxdna_drm_alloc_dev_bo(dev, args, filp);
+		abo = amdxdna_drm_alloc_dev_bo(dev, args, filp, false);
 		break;
 	case AMDXDNA_BO_CMD:
 		abo = amdxdna_drm_create_cmd_bo(dev, args, filp);
 		break;
 	default:
-		ret = -EINVAL;
-		goto err_unlock;
+		return -EINVAL;
 	}
-	if (IS_ERR(abo)) {
-		ret = PTR_ERR(abo);
-		goto err_unlock;
+	if (IS_ERR(abo))
+		return PTR_ERR(abo);
+
+	if (args->type != AMDXDNA_BO_DEV_HEAP) {
+		/* ready to publish object to userspace */
+		ret = drm_gem_handle_create(filp, to_gobj(abo), &args->handle);
+		if (ret)
+			XDNA_ERR(client->xdna, "Create handle failed");
 	}
```

Comment:
Let's put drm_gem_handle_create() inside each amdxdna_drm_alloc_xxx_bo(). It is just one function call, not much of dup code. Any special case in the code needs careful thought since it might be missed next time when you move code around.

Reply:
As discussed offline, we want to keep handle create in the ioctl(). If other part of driver needs to create a object, then just keep object pointer and reference.